### PR TITLE
ADD: staking pallet

### PR DESCRIPTION
**Pull request by coderfengyun**
{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [x] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] New and existing benchmarks pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream modules
* [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pallets/stake/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+ * 1_000_000 * (1 - 1/2) = x - x * (1 / 2^157)
+ *
+ * so x = 500_000
+ *
+ *
+ * In Chinese
+ * 根据等比数列求和公式，求解一下等比数列的和
+ * 1. 设 x 为第一周释放量
+ * 2. 三年总共365/7 * 3 = 156周
+ * 3. 按照3年释放100W币来做归一化，方便各中值的计算
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:65]: 4. 每周减半？

File: pallets/stake/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+ *
+ * x + x/2 + x / 2^2 + ... + x/2^156 = 1_000_000
+ *
+ * 根据等比数列规律: (1 - q) * S = a1 - a_n+1，即S = (a1 - a_n+1) / 1 - q
+ *
+ * 带入该公式：1_000_000 = (x - x/2^157) / (1 - 1/2) -> x = 500_000 / (1 - 1/2^157) ~ 500_000
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs]: ~=

File: pallets/stake/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+ * so x = 500_000
+ *
+ *
+ * In Chinese
+ * 根据等比数列求和公式，求解一下等比数列的和
+ * 1. 设 x 为第一周释放量
+ * 2. 三年总共365/7 * 3 = 156周
+ * 3. 按照3年释放100W币来做归一化，方便各中值的计算
+ *
+ * x + x/2 + x / 2^2 + ... + x/2^156 = 1_000_000
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:68]: 感觉远不到156周就没精度了

Comment (ruibinx) [reply pallets/stake/src/lib.rs:68]: log2 (500000 * (1E18)) = 78
log2 500000 = 18
大概18周之后就少于1Ad3了

File: pallets/stake/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+            rewards[msg.sender] = 0;
+        }else{
+            rewards[msg.sender] = rewards[msg.sender].add(
+                earnings_per_share.mul(amount)
+            );
+        }
+        super.stake(amount);
+        emit Staked(msg.sender, amount);
+         */
+        pub fn stake(
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:355]: 用户stake的时候不需要transfer一笔钱吗？

File: pallets/stake/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+            let per_block_output = Self::get_per_block_output(activity.asset_id)?;
+            let profit = per_block_output.saturating_mul(diff.into());
+            Ok(profit)
+        }
+
+        /**
+        * modifier make_profit() {
+               uint256 amount = getprofit();
+               if (amount > 0) {
+                   yfi.mint(address(this), amount);
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:294]: mint这句似乎没体现？这是在尝试发奖吗？

File: pallets/stake/src/lib.rs (hunk 4 above)
Comment (coderfengyun) [reply pallets/stake/src/lib.rs:355]: 不需要，当前是用liquidity stake

File: pallets/stake/src/lib.rs (hunk 3 above)
Comment (coderfengyun) [reply pallets/stake/src/lib.rs:68]: YFI就是这个pattern, 我们可以考虑改比例，但我感觉意义不大

File: pallets/stake/src/lib.rs (hunk 5 above)
Comment (coderfengyun) [reply pallets/stake/src/lib.rs:294]: 我们不需要，这一句的意思是把币发给当前合约，我们是直接在发reward的时候mint币

File: pallets/stake/src/lib.rs (hunk 3 above)
Comment (ruibinx) [reply pallets/stake/src/lib.rs:68]: 如果三年发完是个硬需求的话，感觉应该改减半周期

File: pallets/stake/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+                return 0;
+            }
+            uint256 diff = new_blocknum.sub(lastblock);
+            lastblock = new_blocknum;
+            uint256 profit = diff.mul(getPerBlockOutput());
+            return profit;
+        }
+         */
+        fn get_profit(activity: &StakingActivityOf<T>) -> Result<BalanceOf<T>, DispatchError> {
+            let cur_block_num = <frame_system::Pallet<T>>::block_number();
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:261]: 如果get_profit两次调用间隔大于等于2倍HalveTime会有问题吗？

Comment (coderfengyun) [reply pallets/stake/src/lib.rs:261]: 会有问题，在make_profit哪里做了防御，取了个min:

`            let amount = amount.min(activity.reward_total_remains);`

File: pallets/swap/src/migrations.rs [hunk 7]
```diff
@@ -107,3 +107,127 @@ pub mod v2 {
...
+                if liquidity_owner != initial_liquidity_provider {
+                    lp_token_count_to_burn += 1;
+                }
+            }
+
+            info!("there are {:} lp_token to burn", lp_token_count_to_burn);
+            Ok(())
+        }
+
+        // #[cfg(feature = "try-runtime")]
```
Comment (ruibinx) [inline pallets/swap/src/migrations.rs]:
```suggestion
        #[cfg(feature = "try-runtime")]
```

File: runtimes/dana/Cargo.toml [hunk 8]
```diff
@@ -186,9 +188,10 @@ runtime-benchmarks = [
     'parami-nft/runtime-benchmarks',
     'parami-swap/runtime-benchmarks',
     'parami-tag/runtime-benchmarks',
+    'parami-stake/runtime-benchmarks',
 
     'sp-runtime/runtime-benchmarks',
-
+
```
Comment (ruibinx) [inline runtimes/dana/Cargo.toml:194]:
```suggestion

```

File: pallets/stake/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+
+        /// The pallet id, used for deriving "pot" accounts of staking activity's reward
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        #[pallet::constant]
+        type OneMillionNormalizedInitDailyOutput: Get<BalanceOf<Self>>;
+
+        #[pallet::constant]
+        type DurationInBlockNum: Get<Self::BlockNumber>;
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:111]: HalveDuration?

File: pallets/stake/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+        type Assets: FungMeta<AccountOf<Self>, AssetId = AssetIdOf<Self>>
+            + FungMutate<AccountOf<Self>, AssetId = Self::AssetId, Balance = BalanceOf<Self>>
+            + FungTransfer<AccountOf<Self>, AssetId = AssetIdOf<Self>, Balance = BalanceOf<Self>>;
+
+        /// The pallet id, used for deriving "pot" accounts of staking activity's reward
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        #[pallet::constant]
+        type OneMillionNormalizedInitDailyOutput: Get<BalanceOf<Self>>;
```
Comment (ruibinx) [inline pallets/stake/src/lib.rs:108]: InitialDailyOutput?

File: pallets/stake/src/lib.rs (hunk 6 above)
Comment (ruibinx) [reply pallets/stake/src/lib.rs:261]: 没太明白，需要面聊一下

File: pallets/stake/src/lib.rs (hunk 10 above)
Comment (coderfengyun) [reply pallets/stake/src/lib.rs:108]: 这个的意思是按照100W的总量做归一化之后的InitDailyOutput

**Pull request merged**